### PR TITLE
Fix hover tabbing behaviour

### PR DESCRIPTION
**Pull request by rzhao271**
Ref #159088

Previously, when pressing the escape key from a focused hover, the user loses the previously focused element and has to tab back to it. This PR attempts to fix that scenario by placing the focus back to the previously focused element.

### CODE REVIEW

**Comment by rzhao271** [comment]
Demo showing tab looping behaviour
Screencap of the Modified elsewhere indicator in the Settings editor being focused. Pressing space opens the hover. Tabbing around goes to the various links. Pressing escape closes the hover and focuses back on the indicator.

File: src/vs/workbench/services/hover/browser/hoverService.ts [hunk 1]
```diff
@@ -23,6 +23,8 @@ export class HoverService implements IHoverService {
 	private _currentHoverOptions: IHoverOptions | undefined;
 	private _currentHover: HoverWidget | undefined;
 
+	private _lastFocusedElementBeforeOpen: HTMLElement | null = null;
```
Comment (Tyriar) [inline src/vs/workbench/services/hover/browser/hoverService.ts:26]: Let's use undefined if there are only 2 states

File: src/vs/workbench/services/hover/browser/hoverWidget.ts [hunk 2]
```diff
@@ -46,12 +46,15 @@ export class HoverWidget extends Widget {
 	private readonly _target: IHoverTarget;
 	private readonly _linkHandler: (url: string) => any;
 
+	private readonly _lastFocusedElementBeforeOpen: HTMLElement | null;
```
Comment (Tyriar) [inline src/vs/workbench/services/hover/browser/hoverWidget.ts:49]: Do we need to track this element in both the service and the widget?

File: src/vs/workbench/services/hover/browser/hoverWidget.ts [hunk 3]
```diff
@@ -46,12 +46,15 @@ export class HoverWidget extends Widget {
...
+
 	private _isDisposed: boolean = false;
 	private _hoverPosition: HoverPosition;
 	private _forcePosition: boolean = false;
 	private _x: number = 0;
 	private _y: number = 0;
 	private _isLocked: boolean = false;
+	private _addedTabLoop: boolean = false;
```
Comment (Tyriar) [inline src/vs/workbench/services/hover/browser/hoverWidget.ts:57]: I think "focus trap" is a more common name for this

File: src/vs/workbench/services/hover/browser/hoverWidget.ts [hunk 4]
```diff
@@ -222,10 +228,53 @@ export class HoverWidget extends Widget {
 		}
 	}
 
+	private addTabLoop() {
+		// Add a hover tab loop if the hover has at least one element with a valid tabIndex
+		const firstContainerFocusElement = this._hover.containerDomNode;
+		const lastContainerFocusElement = this.findLastFocusableChild(this._hover.contentsDomNode);
+		if (lastContainerFocusElement) {
+			const beforeContainerFocusElement = dom.prepend(this._hoverContainer, $('div.focus-before'));
+			const afterContainerFocusElement = dom.append(this._hoverContainer, $('div.focus-after'));
```
Comment (Tyriar) [inline src/vs/workbench/services/hover/browser/hoverWidget.ts:237]: I don't think these classes are used

File: src/vs/workbench/services/hover/browser/hoverWidget.ts [hunk 5]
```diff
@@ -222,10 +228,53 @@ export class HoverWidget extends Widget {
...
+			}));
+		}
+	}
+
+	private findLastFocusableChild(root: Node): HTMLElement | undefined {
+		if (root.hasChildNodes()) {
+			for (let i = 0; i < root.childNodes.length; i++) {
+				const node = root.childNodes.item(root.childNodes.length - i - 1);
+				if (node.nodeType === node.ELEMENT_NODE) {
+					const parsedNode = <HTMLElement>node;
```
Comment (Tyriar) [inline src/vs/workbench/services/hover/browser/hoverWidget.ts:256]: nit

```suggestion
					const parsedNode = node as HTMLElement;
```

File: src/vs/workbench/services/hover/browser/hoverWidget.ts [hunk 6]
```diff
@@ -222,10 +228,53 @@ export class HoverWidget extends Widget {
...
+	}
+
 	public render(container: HTMLElement): void {
 		container.appendChild(this._hoverContainer);
 
 		this.layout();
+		if (!this._addedTabLoop) {
+			this.addTabLoop();
+			this._addedTabLoop = true;
+		}
```
Comment (Tyriar) [inline src/vs/workbench/services/hover/browser/hoverWidget.ts:277]: This would be cleaner/safer moved into `addTabLoop`

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/services/hover/browser/hoverService.ts [hunk 7]
```diff
@@ -23,6 +23,8 @@ export class HoverService implements IHoverService {
 	private _currentHoverOptions: IHoverOptions | undefined;
 	private _currentHover: HoverWidget | undefined;
```
Comment (alexr00) [inline src/vs/workbench/services/hover/browser/hoverService.ts:25]: <span style="background-color:#542426;">&nbsp;&nbsp;Hello&nbsp;&nbsp;</span>

**Review by Tyriar — APPROVED** [review]

**Pull request merged**